src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -1398,6 +1420,75 @@ public Type getReturnType(String invocationId) {
         }
     }
 
+    // We don't have reconnect yet, but this helps align the Java client with the .NET client
+    // and hopefully make it easier to implement reconnect in the future
+    private final class ReconnectingConnectionState {
+        private final Logger logger;
+        private final Lock lock = new ReentrantLock();
+        private ConnectionState state;
+        private HubConnectionState hubConnectionState = HubConnectionState.DISCONNECTED;
+        public Completable startTask;
+        public TransportEnum currentTransport;
+
+        public ReconnectingConnectionState(Logger logger) {
+            this.logger = logger;
+        }
+
+        public void setConnectionState(ConnectionState state) {
+            this.lock.lock();
+            try {
+                this.state = state;
+            } finally {
+                this.lock.unlock();
+            }
+        }
+
+        public ConnectionState getConnectionStateUnsynchronized() {
+            return this.state;
```

Comment:
I know this is always called with a lock (except for getConnectionId()), but a lot of the places where it's used don't null check the result. I think we should throw a "Connection is not active." exception when state is null here by default and have a special version of this method (or use a flag) so getConnectionId() can skip the null check.